finatra-http: Increase composability and flexibility of RouteDSL

chrisbenincasa opened the pull request:
Problem:

Today, the routing DSL supports two composable pieces (prefix and
filter). These are implemented under different paradigms; filter
produces a new Route context while prefix takes a block and uses a var
to track the prefix.

There are a few issues / unexpected behaviors that result:
* Prefixes are not composable
* Filtered routes inside of prefixes don't work (https://github.com/twitter/finatra/issues/392)
* Filtered prefixes are not supported
* The syntax is rigid between the two

Solution

The changes here revist the structure of RouteDSL and unify the way
routes are defined in Controllers to increase flexibility and remove
unexpected behavior and bugs. Namely, mutable state is now centralized
to a twitter-util Var cell and an API is laid out for creating
"specialized" RouteDSL contexts (FilteredDSL and PrefixedDSL).

This allows for a great amount of flexibility for the user when defining
routes as well as increasing the extensibility of the DSL for future
additions. For example:

Filtered prefixes:
```scala
filter[MyFilter].prefix("/v1/prefix") {
    // ...
}
```

Cascading prefixes:
```scala
prefix("/v1") {
  prefix("/foo") {
    prefix("/bar") {
     // ..
    }
  }
}
```

"Freestyle" filter syntax:

```scala
filter[MyFilter] {
  // ...
}
```

cacoco commented on http/src/main/scala/com/twitter/finatra/http/RouteDSL.scala:
```diff
@@ -4,35 +4,70 @@ import com.twitter.finagle.Filter
+private[http] case class RouteContext(prefix: String, buildFilter: (Injector) => HttpFilter)
 
+private trait MutableRouteState {
+  private[http] val context: RouteContext = RouteContext("", (_) => Filter.identity)
```
can you name the params here? e.g.,

```
... = RouteContext(
  prefix = "",
  buildFilter = (_) => Filter.identity
```

cacoco commented on http/src/main/scala/com/twitter/finatra/http/RouteDSL.scala:
```diff
@@ -99,6 +134,19 @@ private[http] trait RouteDSL { self =>
+  /* Protected */
+
+  // A function that wraps another and sets any contexts, if necessary
+  protected def contextWrapper[T](f: => T): T = f
```
I would prefer that this is has a tighter visibility in that it's only used in the RouteDSL as we don't want to expose it for general use. And anything not locked down will be used making it harder to refactor later.

chrisbenincasa replied:
Totally. Will fix. Is `private[http]` ok?

cacoco replied:
That should work but if we don't need to leak it to the package, e.g. `private` or `private[this]` that would be even more preferable (same with the other usages of `private[http]` throughout).

chrisbenincasa replied:
I think `private[http]` is the tightest we can use here, since these methods are overridden by `PrefixedDSL` and `FilteredDSL`

cacoco commented on http/src/main/scala/com/twitter/finatra/http/RouteDSL.scala:
```diff
@@ -99,6 +134,19 @@ private[http] trait RouteDSL { self =>
+  protected def contextWrapper[T](f: => T): T = f
+
+  // Executes a block with a given RouteContext
+  protected def withContext[T](ctx: RouteContext)(f: => T): T = {
```
same as above.

chrisbenincasa replied:
👍 same with `private[http]` here. This method is referenced in the subclasses, so I think package private is the tightest we can get.

lou-k commented on http/src/main/scala/com/twitter/finatra/http/RouteBuilder.scala:
```diff
@@ -1,9 +1,10 @@
 package com.twitter.finatra.http
 
-import com.twitter.finagle.http.{Method => HttpMethod, RouteIndex}
+import com.twitter.finagle.http.{RouteIndex, Method => HttpMethod}
```
This diff seems un-necessary.

lou-k commented on http/src/main/scala/com/twitter/finatra/http/RouteDSL.scala:
```diff
@@ -99,6 +134,19 @@ private[http] trait RouteDSL { self =>
+  protected def withContext[T](ctx: RouteContext)(f: => T): T = {
+    val orig = contextRef()
+    contextRef() = ctx
+    val res = try f finally contextRef() = orig
```
seems like you can drop this `val`

chrisbenincasa commented:
@cacoco Made the suggested changes and re-pushed. Let me know what you think!

cacoco commented:
@chrisbenincasa I'll take a look at seeing how well this merges internally. I may have more updates for you pending that work and more internal reviews. Thanks!

cacoco commented on http/src/test/scala/com/twitter/finatra/http/tests/integration/doeverything/main/controllers/DoEverythingController.scala:
```diff
@@ -62,6 +62,38 @@ class DoEverythingController @Inject()(
+
+    filter(new AppendToHeaderFilter("appended" , "1")).
+      filter(new AppendToHeaderFilter("appended" , "2")).
+      get("/appendMultiplePrefixed") { request: Request =>
```
there's no prefix call here, is that intentional? Seems like this was meant to test filter and prefix together?

chrisbenincasa replied:
Yes, this was intentional. A side effect of reworking how contexts work meant the DSL now permits this syntax (which I think has its own use cases and gives users some more flexibility). Included the test here as an example / get coverage.

cacoco replied:
@chrisbenincasa I'm not sure I understand. Users could always call `filter.filter.get` before and the route says it's "appendMultiplePrefixed" but there's no prefix call? So, at a minimum maybe the route needs to be changed as it's confusing?

chrisbenincasa replied:
@cacoco apologies, I may have misunderstood your question. This _is_ within a prefix block. See [here](https://github.com/twitter/finatra/pull/393/files/d662153cf678cac709526ca0b388961ae9fddd04#diff-3ed2e8ed4110fe9e36488404995d1498R57). I stuck this along with the existing prefix tests.  

I misread my own diff here 😆 I thought we were discussing the "freestyle" syntax of filters that is now supported as a side-effect (like [in this test](https://github.com/twitter/finatra/pull/393/files/d662153cf678cac709526ca0b388961ae9fddd04#diff-3ed2e8ed4110fe9e36488404995d1498R76))

cacoco replied:
@chrisbenincasa Ah! I see. It's way above, got it. Thanks for the clarification.

cacoco commented on http/src/main/scala/com/twitter/finatra/http/RouteDSL.scala:
```diff
@@ -4,35 +4,74 @@ import com.twitter.finagle.Filter
 import scala.collection.mutable.ArrayBuffer
 
-private[http] trait RouteDSL { self =>
+private[http] case class RouteContext(prefix: String, buildFilter: (Injector) => HttpFilter)
```
Can you comment this class? It would really help.

chrisbenincasa replied:
Definitely.

cacoco commented on http/src/main/scala/com/twitter/finatra/http/RouteDSL.scala:
```diff
@@ -4,35 +4,74 @@ import com.twitter.finagle.Filter
+private trait MutableRouteState {
+  private[http] val context: RouteContext = RouteContext(
+    prefix = "",
+    buildFilter = (_) => Filter.identity
```
ffti: Function.const(Filter.identity)

cacoco commented on http/src/main/scala/com/twitter/finatra/http/RouteDSL.scala:
```diff
@@ -4,35 +4,74 @@ import com.twitter.finagle.Filter
+    buildFilter = (_) => Filter.identity
+  )
+
+  private[http] lazy val contextRef = Var(context)
```
An explanation for why this is lazy would be good, considering the value stored in it is updatable.

chrisbenincasa replied:
Absolutely. I made this lazy to protect against initialization order issues.

For instance, if we remove `lazy` from `contextRef`, Controller initialization will fail with a NPE. This is because we are evaluating the bottom of the "hierarchy" first. 

So, in this example, `ReadHeadersController` starts initializing, calls `filter` which creates a new FilteredDSL referencing the parent's `contextRef` and eagerly evaluating the `context` (reliant on `contextRef`) which has not been initialized yet (because we haven't yet reached `MutableRouteState` in the linearization of `ReadHeadersController`). 

I can drop a comment about this in the code as well, if you wish.

cacoco replied:
@chrisbenincasa yes, please.

cacoco commented on http/src/main/scala/com/twitter/finatra/http/RouteDSL.scala:
```diff
@@ -4,35 +4,74 @@ import com.twitter.finagle.Filter
+  private[http] lazy val contextRef = Var(context)
+}
+
+private class FilteredDSL[FilterType <: HttpFilter : Manifest] extends RouteDSL {
```
Also private[http] and please add tests?

chrisbenincasa replied:
Will do.

cacoco commented on http/src/main/scala/com/twitter/finatra/http/RouteDSL.scala:
```diff
@@ -4,35 +4,74 @@ import com.twitter.finagle.Filter
+  override private[http] def contextWrapper[T](f: => T): T = withContext(context)(f)
+}
+
+private class PrefixedDSL(prefix: String) extends RouteDSL {
```
Also private[http] and please add tests?

chrisbenincasa commented:
@cacoco just pushed up changes into a new commit. Added some comments like you suggested and did a minor refactor on the the default value of `context` to break the val dependency between `contextRef` and `context`. Also added tests for FilteredDSL and PrefixedDSL in isolation and used together.

cacoco commented:
@chrisbenincasa great, we'll take a look!

cacoco commented:
Merged in 49957f12dc10f7a3abf909c09e5fb93bcbc9ee19
